### PR TITLE
WASM On Ray Support REP

### PR DESCRIPTION
**Pull request by wilsonwang371**
This is a pull request for adding wasm support to Ray.

### CODE REVIEW

**Comment by wilsonwang371** [comment]
@iycheng @scv119

**Comment by juntao** [comment]
This is great! I would like to add that, while there are challenges in supporting native AI/ML frameworks in WASM, there are successful use cases and even standards to accomplish this. The [WASI-NN](https://github.com/WebAssembly/wasi-nn) standard allows WASM bytecode programs to access ML frameworks. Tensorflow, Tensorflow Lite, PyTorch, and OpenVINO are all supported. There are Rust APIs for this:

https://wasmedge.org/book/en/write_wasm/rust/wasinn.html

and JavaScript APIs:

https://wasmedge.org/book/en/write_wasm/js/tensorflow.html

Would like to collaborate and see Ray libraries supported in WASM!

File: reps/2023-02-01-wasm-on-ray.md
Comment (iycheng) [inline reps/2023-02-01-wasm-on-ray.md:11]: This seems like a big adds to WASM if Ray can have WASM workers, in summary, seems like:

Ray adds on WASM:
- ObjectRef for communication
- Ray worker for parallelism and scale up

WASM adds on Ray:
- Lightweight x-lang support
- Lightweight isolation

Comment (iycheng) [reply reps/2023-02-01-wasm-on-ray.md:11]: cc @scv119

**Comment by iycheng** [comment]
Synced it offline and I think the motivation is much clearer. I think the MVP features needed for this use case should be briefly discussed so for the people who don't understand WASM (or Ray), they know the work in high-level.

File: reps/2023-02-01-wasm-on-ray.md
Comment (iycheng) [inline reps/2023-02-01-wasm-on-ray.md:19]: I think for Ray Serve ML inference, the ML also runs in the memory. Could you give more detail about the difference between going with Ray Serve and WASI-NN ?

File: reps/2023-02-01-wasm-on-ray.md
Comment (jjyao) [inline reps/2023-02-01-wasm-on-ray.md:158]: It's a significant con.

Comment (wilsonwang371) [reply reps/2023-02-01-wasm-on-ray.md:158]: I think this part can help us review some of the Ray core APIs so that we can improve the Ray core interface. For the MVP, I think we cannot yet support Actor. With more brainstorming and design sessions, I think there will be a solution. I didn't see any technical challenges so far.

File: reps/2023-02-01-wasm-on-ray.md
Comment (wilsonwang371) [reply reps/2023-02-01-wasm-on-ray.md:19]: wasi-nn is a low-level ML computing API. https://github.com/WebAssembly/wasi-nn/blob/main/wasi-nn.wit.md such as loading a model, setting input, computing with the model, and getting the output. In contrast, Ray Serve is a high-level Python API for the model serving. In order to do something similar in WASM, we need to port some of the Ray Serve from the Python worker to the WASM worker.

File: reps/2023-02-01-wasm-on-ray.md
Comment (iycheng) [reply reps/2023-02-01-wasm-on-ray.md:158]: @jjyao there are several things here:

- Build workers
    - support functions (first step)
    - support actors (third step)
- Define middle layer apis (second step)

The difficulty comes from ray side. To be more specific, ray doesn't have good developer api to add new languages. And this is one of the goal of this project, to figure out what might be good api we should define.

Comment (jjyao) [reply reps/2023-02-01-wasm-on-ray.md:158]: Figuring out a good middle layer for multi-language support is indeed very important (IMO, more important than supporting WASM itself), are we going to have a design for that?

Comment (iycheng) [reply reps/2023-02-01-wasm-on-ray.md:158]: @jjyao I know what you mean. The fact here is that WASM worker support is critical to @wilsonwang371 and middle layer support is important for Ray. Given that we don't know what's needed, it's a good chance we'll learn it from this work by doing first step (support functions).

TBH, depend on the resources we have it might or might not happen. But at lease we'll understand the requirement better after finishing step 1.

Comment (wilsonwang371) [reply reps/2023-02-01-wasm-on-ray.md:158]: totally agreed. Actually, we are also willing to contribute to the middle layer API.

File: reps/2023-02-01-wasm-on-ray.md
Comment (jovany-wang) [inline reps/2023-02-01-wasm-on-ray.md:158]: @iycheng @wilsonwang371 
From your design, if we don't re-design a better lower level API for developers, it's not easy to implement WASM actor, right?
I'm concerning that it's verbose to do all these stuffs in one.
We should separate this REP to 2 proposals: a better lower level API for multi-languages workers proposal + WASM worker proposal.

The benefits are significant:
1. For the lower level API proposal, I guess there're many committers and reviewers want to review, while that the WASM worker proposal is focusing on how we port WASM on Ray, which I believe less of reviewers is available to review due to it requires more knowledge about WASM(For example, I'm lack of that).
2. Different goals should use different REPs for better track in the future.

**Review by jovany-wang — CHANGES_REQUESTED** [review]

File: reps/2023-02-01-wasm-on-ray.md
Comment (wilsonwang371) [reply reps/2023-02-01-wasm-on-ray.md:158]: I think WASM workers will not require a better lower-level API to be there in the first place. But our Bytedance team is definitely willing to work together on the lower-level API design & implementation.

Comment (wilsonwang371) [reply reps/2023-02-01-wasm-on-ray.md:158]: Besides, the lower-level API is not a part of this REP.

Comment (jovany-wang) [reply reps/2023-02-01-wasm-on-ray.md:158]: I see your point that this REP doesn't include the lower level API here.

I'm worrying about whether there might be a risk, like once this REP getting merged, and then it's impossible to support WASM actor.

If you guys are confident for this, let us make it clearer.

Comment (wilsonwang371) [reply reps/2023-02-01-wasm-on-ray.md:158]: there will be ways to support actor in wasm. But we need to choose a solution that is acceptable to most people. That's why this is not a part of the MVP version

Comment (jovany-wang) [reply reps/2023-02-01-wasm-on-ray.md:158]: OK. Please let me know once you have design or some stuffs. Looking forward the contributions.

Comment (iycheng) [reply reps/2023-02-01-wasm-on-ray.md:158]: I think the issues here are that no one knows how to do it :(

Before starting actor support, we probably should get to the api layer first and we'll feel comfortable about this at that stage.

**Review by iycheng — APPROVED** [review]

**Comment by zhe-thoughts** [comment]
Given the vote on ray-committers list, merging this PR. Thanks for the contribution @wilsonwang371 and @iycheng for shepherding!

**Pull request merged**
